File: src/main/java/xf/xflp/base/position/TouchingPerimeterService.java
package xf.xflp.base.position;

import xf.xflp.base.container.Container;
import xf.xflp.base.item.Item;
import xf.xflp.base.item.Position;

import java.util.List;

/**
 * Copyright (c) 2012-2026 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @author hschneid
 */
public class TouchingPerimeterService {

    public static float getTouchingPerimeter(
            Container container,
            PositionCandidate candidate,
            int itemTouchValue,
            boolean considerWalls,
            boolean considerBaseFloor) {
        Item item = candidate.item();
        Position pos = candidate.position();

        int value = 0;
        int w = item.w;
        int l = item.l;
        int h = item.h;
        if(candidate.isRotated()) {
            w = item.l;
            l = item.w;
        }

        int posX = pos.x();
        int posY = pos.y();
        int posZ = pos.z();
        int xw = posX + w;
        int yl = posY + l;
        int zh = posZ + h;

        List<Integer> positionCoords, itemCoords;

        // x-Achse
        {
            if(posX == 0 && considerWalls)
                value += h * l;
            if(xw == container.getWidth() && considerWalls)
                value += h * l;

            positionCoords = container.getBaseData().getXMap().get(posX);
            itemCoords = container.getBaseData().getXMap().get(xw);

            value += touchValueX(positionCoords, container, posX, xw, posY, yl, posZ, zh, itemTouchValue);
            value += touchValueX(itemCoords, container, posX, xw, posY, yl, posZ, zh, itemTouchValue);
        }

        // Y-Achse
        {
            if(posY == 0 && considerWalls)
                value += h * w;
            if(yl == container.getLength() && considerWalls)
                value += h * w;

            positionCoords = container.getBaseData().getYMap().get(posY);
            itemCoords = container.getBaseData().getYMap().get(yl);

            value += touchValueY(positionCoords, container, posX, xw, posY, yl, posZ, zh, itemTouchValue);
            value += touchValueY(itemCoords, container, posX, xw, posY, yl, posZ, zh, itemTouchValue);
        }

        // Z-Achse
        {
            if(posZ == 0 && considerBaseFloor)
                value += w * l;
            if(zh == container.getHeight() && considerWalls)
                value += w * l;

            positionCoords = container.getBaseData().getZMap().get(posZ);
            itemCoords = container.getBaseData().getZMap().get(zh);

            value += touchValueZ(positionCoords, container, posX, xw, posY, yl, posZ, zh, itemTouchValue);
            value += touchValueZ(itemCoords, container, posX, xw, posY, yl, posZ, zh, itemTouchValue);
        }

        return value;
    }

    private static int touchValueX(List<Integer> itemIndices, Container container,
                                    int posX, int xw, int posY, int yl, int posZ, int zh,
                                    int itemTouchValue) {
        if (itemIndices == null) return 0;
        int val = 0;
        for (int j = itemIndices.size() - 1; j >= 0; j--) {
            Item i = container.getItems().get(itemIndices.get(j));
            if (i.xw == posX || i.x == xw) {
                if (i.y > yl || i.yl < posY) continue;
                if (i.z > zh || i.zh < posZ) continue;
                int yLength = Math.min(yl, i.yl) - Math.max(i.y, posY);
                int zLength = Math.min(zh, i.zh) - Math.max(i.z, posZ);
                val += yLength * zLength * itemTouchValue;
            }
        }
        return val;
    }

    private static int touchValueY(List<Integer> itemIndices, Container container,
                                    int posX, int xw, int posY, int yl, int posZ, int zh,
                                    int itemTouchValue) {
        if (itemIndices == null) return 0;
        int val = 0;
        for (int j = itemIndices.size() - 1; j >= 0; j--) {
            Item i = container.getItems().get(itemIndices.get(j));
            if (i.yl == posY || i.y == yl) {
                if (i.x > xw || i.xw < posX) continue;
                if (i.z > zh || i.zh < posZ) continue;
                int xLength = Math.min(xw, i.xw) - Math.max(i.x, posX);
                int zLength = Math.min(zh, i.zh) - Math.max(i.z, posZ);
                val += xLength * zLength * itemTouchValue;
            }
        }
        return val;
    }

    private static int touchValueZ(List<Integer> itemIndices, Container container,
                                    int posX, int xw, int posY, int yl, int posZ, int zh,
                                    int itemTouchValue) {
        if (itemIndices == null) return 0;
        int val = 0;
        for (int j = itemIndices.size() - 1; j >= 0; j--) {
            Item i = container.getItems().get(itemIndices.get(j));
            if (i.zh == posZ || i.z == zh) {
                if (i.y > yl || i.yl < posY) continue;
                if (i.x > xw || i.xw < posX) continue;
                int yLength = Math.min(yl, i.yl) - Math.max(i.y, posY);
                int xLength = Math.min(xw, i.xw) - Math.max(i.x, posX);
                val += yLength * xLength * itemTouchValue;
            }
        }
        return val;
    }
}

File: src/xflp/base/position/touching_perimeter_service.py
"""
Copyright (c) 2012-2025 Holger Schneider
All rights reserved.

This source code is licensed under the MIT License (MIT) found in the
LICENSE file in the root directory of this source tree.
"""

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from xflp.base.container import Container
    from .position_candidate import PositionCandidate


class TouchingPerimeterService:
    """Service for calculating touching perimeter of items."""

    @staticmethod
    def get_touching_perimeter(container: 'Container',
                              candidate: 'PositionCandidate',
                              item_touch_value: int,
                              consider_walls: bool,
                              consider_base_floor: bool) -> float:
        """
        Calculate the touching perimeter for an item at a candidate position.

        Args:
            container: The container
            candidate: The position candidate
            item_touch_value: Value multiplier for item touching
            consider_walls: Whether to consider container walls
            consider_base_floor: Whether to consider base floor

        Returns:
            The touching perimeter value
        """
        item = candidate.item
        pos = candidate.position

        value = 0
        w = item.w
        l = item.l
        h = item.h

        if candidate.is_rotated:
            w = item.l
            l = item.w

        xw = pos.x + w
        yl = pos.y + l
        zh = pos.z + h

        # X-Axis
        if pos.x == 0:
            # If walls must be considered, full side area is added
            if consider_walls:
                value += h * l

        if xw == container.get_width():
            # If walls must be considered, full side area is added
            if consider_walls:
                value += h * l

        x_item_list = []
        x_map = container.get_base_data().get_x_map()
        if pos.x in x_map:
            x_item_list.extend(x_map[pos.x])
        if xw in x_map:
            x_item_list.extend(x_map[xw])

        if len(x_item_list) > 0:
            # Check all items which touch pos.x
            for j in range(len(x_item_list) - 1, -1, -1):
                index = x_item_list[j]
                i = container.get_items()[index]

                if i.xw == pos.x or i.x == xw:
                    # Check length and height
                    if i.y > yl or i.yl < pos.y:
                        continue
                    if i.z > zh or i.zh < pos.z:
                        continue

                    # If items touch themselves, calculate the cutting plane
                    y_length = min(yl, i.yl) - max(i.y, pos.y)
                    z_length = min(zh, i.zh) - max(i.z, pos.z)
                    value += y_length * z_length * item_touch_value

        # Y-Axis
        if pos.y == 0:
            # If walls must be considered, full side area is added
            if consider_walls:
                value += h * w

        if yl == container.get_length():
            # If walls must be considered, full side area is added
            if consider_walls:
                value += h * w

        y_item_list = []
        y_map = container.get_base_data().get_y_map()
        if pos.y in y_map:
            y_item_list.extend(y_map[pos.y])
        if yl in y_map:
            y_item_list.extend(y_map[yl])

        if len(y_item_list) > 0:
            # Check all items which touch pos.y
            for j in range(len(y_item_list) - 1, -1, -1):
                index = y_item_list[j]
                i = container.get_items()[index]

                if i.yl == pos.y or i.y == yl:
                    # Check width and height
                    if i.x > xw or i.xw < pos.x:
                        continue
                    if i.z > zh or i.zh < pos.z:
                        continue

                    # If items touch themselves, calculate the cutting plane
                    x_length = min(xw, i.xw) - max(i.x, pos.x)
                    z_length = min(zh, i.zh) - max(i.z, pos.z)
                    value += x_length * z_length * item_touch_value

        # Z-Axis
        if pos.z == 0:
            # If walls must be considered, full side area is added
            if consider_base_floor:
                value += w * l

        if zh == container.get_height():
            # If walls must be considered, full side area is added
            if consider_walls:
                value += w * l

        z_item_list = []
        z_map = container.get_base_data().get_z_map()
        if pos.z in z_map:
            z_item_list.extend(z_map[pos.z])
        if zh in z_map:
            z_item_list.extend(z_map[zh])

        # Check all items which touch pos.z
        for j in range(len(z_item_list) - 1, -1, -1):
            index = z_item_list[j]
            i = container.get_items()[index]

            if i.zh == pos.z or i.z == zh:
                # Check length and width
                if i.y > yl or i.yl < pos.y:
                    continue
                if i.x > xw or i.xw < pos.x:
                    continue

                # If items touch themselves, calculate the cutting plane
                y_length = min(yl, i.yl) - max(i.y, pos.y)
                x_length = min(xw, i.xw) - max(i.x, pos.x)
                value += y_length * x_length * item_touch_value

        return value
